### PR TITLE
Don't synchronize MapperService.doMerge for preflight checks

### PR DESCRIPTION
**Pull request by original-brownbear**
We don't need to synchronize if we are not actually updating the mapping field. This is needlessly slowing down dynamic mapping updates and blocking the write pool.
Also, this moves the mapping parsing outside the synchronized block in all cases, no need to have that run exclusively as well.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -560,18 +560,24 @@ public DocumentMapper merge(String type, CompressedXContent mappingSource, Merge
...
-        DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
         if (reason == MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT) {
-            return newMapper;
+            Mapping mapping = mergeMappings(this.mapper, incomingMapping, MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT, this.indexSettings);
+            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
+            // We should identify these cases reliably and save expensive serialization here
+            return newDocumentMapper(mapping, MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT, mapping.toCompressedXContent());
+        } else {
+            synchronized (this) {
+                Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason, this.indexSettings);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:572]: Maybe de-duplicate the TODO given it's the same, and leave a clarifying comment on the distinction and why mergeMappings needs to be under synchronized in one case but not on the other.

**Review by javanna — APPROVED** [review]
Looks correct to me. I wonder if we should add additional tests to cover this specific scenario. Are you aware of existing ones that already provide coverage for concurrent dynamic mappings updates?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:575]: ++ makes sense, added

**Comment by original-brownbear** [comment]
> Are you aware of existing ones that already provide coverage for concurrent dynamic mappings updates?

So for the pre-flight path yes, this is essentially covered by all kinds of ITs that run the bulk indexer and go through some slightly heterogeneous documents.
For the path that is currently synchronized, I don't think concurrency is possible :P Just looked through this, this is only ever called from a master update thread or from the thread that node startup runs on, that's it. We can remove the second synchronized block and add an assertion in a follow up :) 

Thanks Luca!

**Pull request merged**